File: Econovation_Algorithm_Mentoring/2. 분할정복, 이분탐색(Divide and Conquer)/Programmers_Immigration.java
import java.util.Arrays;

public class Programmers_Immigration {
    public static long solution(int n, int[] times) {
        long answer = 0;
        Arrays.sort(times);
        long left = 0;
        // 이분탐색의 right 부분을 제대로 설정시킬 것
        long right = (long) n * times[times.length - 1];
        long mid = 0;
        long min = Integer.MAX_VALUE;
        // 굳이 따로 min 필요 없이 그대로 mid를 채택하면 된다..

        while (left <= right) {
            mid = (left + right) / 2;

            // 왠만하면 long을 다 취할 것
            long count = 0;
            for (int i = 0; i < times.length; i++) {
                count += mid / times[i];
            }

            if (count >= n) {
                right = mid - 1;
                min = mid;
            } else {
                left = mid + 1;
            }
        }
        answer = min;
        return answer;
    }

    public static void main(String[] args) {
        int[] arr = {7, 10};
        System.out.println(solution(6, arr));
    }
}

File: Algorithm_Swift/Algorithm_Swift/Programmers/Programmers_Immigration.swift
//
//  Programmers_Immigration.swift
//  Algorithm_Swift
//
//  Created by asong on 2022/04/21.
//

import Foundation

func solutionImmigration(_ n:Int, _ times:[Int]) -> Int64 {
    var time = times.sorted(by: <)
    var left: Int64 = 0
    var right: Int64 = Int64(n * time[time.count-1])
    var mid: Int64 = 0
    var min: Int64 = Int64.max
    while left <= right{
        mid = (left + right) / 2
        
        var count = 0
        for i in 0..<time.count{
            count += Int(mid) / time[i]
        }
        if count >= n{
            right = mid - 1
            min = Swift.min(min,mid)
        }else{
            left = mid + 1
        }
    }
    return min
}
